pkg/cleaner/cleaner.go

```diff
@@ -0,0 +1,154 @@
+	}
+
+	err = c.checkCRDsNotExist(c.apiextensionsClient)
+	if err != nil {
+		return fmt.Errorf("failed while checking CRDs not exist: %s", err)
+	}
+
+	return nil
+}
+
+func (c *Cleaner) scaleDownController(namespace, controllerName string) error {
+	klog.V(4).Infof("Fetching deployment %s/%s", namespace, controllerName)
+	deployment, err := c.client.
+		AppsV1beta1().
+		Deployments(namespace).
+		Get(controllerName, v1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("cannot get deployment %s/%s: %s", namespace, controllerName, err)
+	}
+
+	klog.V(4).Info("Scaling down deployment to zero")
+	replicas := int32(0)
+	deploymentCopy := deployment.DeepCopy()
+	deploymentCopy.Spec.Replicas = &replicas
+	_, err = c.client.AppsV1beta1().Deployments(deploymentCopy.Namespace).Update(deploymentCopy)
+	if err != nil {
+		return fmt.Errorf("failed to update deployment %s/%s: %v", namespace, controllerName, err)
+	}
+
+	err = wait.Poll(3*time.Second, 120*time.Second, func() (done bool, err error) {
+		klog.V(4).Info("Waiting for deployment scales down...")
+		deployment, err := c.client.AppsV1beta1().Deployments(namespace).Get(controllerName, v1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		ready := deployment.Status.ReadyReplicas
+		available := deployment.Status.AvailableReplicas
+		if ready == 0 && available == 0 {
+			return true, nil
+		}
+		klog.V(4).Info("Controller manager is not down, (ready: %d, available: %d) retry...", ready, available)
+		return false, nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed during waiting for scale down controller manager: %s", err)
+	}
+
+	return nil
+}
+
+func (c *Cleaner) removeCRDs(apiextensionsClient apiextensionsclientset.Interface) error {
+	klog.V(4).Info("Removing all ServiceCatalog CustomResourceDefinitions")
+	list, err := apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().List(v1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list CustomResourceDefinition: %s", err)
+	}
+	for _, crd := range list.Items {
+		if !probe.IsServiceCatalogCustomResourceDefinition(crd) {
+			continue
+		}
+		err := apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(crd.Name, &v1.DeleteOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to remove CRD %q: %s", crd.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Cleaner) checkCRDsNotExist(apiextensionsClient apiextensionsclientset.Interface) error {
+	klog.V(4).Info("Checking all ServiceCatalog CustomResourceDefinitions are removed")
+	list, err := apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().List(v1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list CustomResourceDefinition: %s", err)
+	}
+	var amount int
+	for _, crd := range list.Items {
+		if probe.IsServiceCatalogCustomResourceDefinition(crd) {
+			amount++
+		}
+	}
+
+	if amount != 0 {
+		return fmt.Errorf("CustomResourceDefinitions list is not empty. There are %d CRD(s)", amount)
```

Comment:
maybe we can add also the name. The easiest way is to just replace the `amount` with slice `[]string` in which you will append name of the CRD that is still not removed